src/realm/table_view.cpp

```diff
@@ -509,41 +504,69 @@ void TableView::do_sort(const DescriptorOrdering& ordering)
-        ObjKey key = get_key(t);
-        if (m_table->is_valid(key)) {
-            index_pairs.emplace_back(key, t);
+    bool using_indexpairs = false;
+
+    auto apply_indexpairs = [&] {
+        m_key_values.clear();
+        for (auto& pair : index_pairs) {
+            m_key_values.add(pair.key_for_object);
         }
-        else
-            ++detached_ref_count;
-    }
+        for (size_t t = 0; t < detached_ref_count; ++t)
+            m_key_values.add(null_key);
+        using_indexpairs = false;
+    };
+
+    auto use_indexpairs = [&] {
+        index_pairs.reserve(sz);
+        index_pairs.clear();
+        // always put any detached refs at the end of the sort
+        // FIXME: reconsider if this is the right thing to do
+        // FIXME: consider specialized implementations in derived classes
+        // (handling detached refs is not required in linkviews)
+        for (size_t t = 0; t < sz; t++) {
+            ObjKey key = get_key(t);
+            if (m_table->is_valid(key)) {
+                index_pairs.emplace_back(key, t);
+            }
+            else
+                ++detached_ref_count;
+        }
+        using_indexpairs = true;
+    };
 
     const int num_descriptors = int(ordering.size());
     for (int desc_ndx = 0; desc_ndx < num_descriptors; ++desc_ndx) {
         const BaseDescriptor* base_descr = ordering[desc_ndx];
         const BaseDescriptor* next = ((desc_ndx + 1) < num_descriptors) ? ordering[desc_ndx + 1] : nullptr;
-        BaseDescriptor::Sorter predicate = base_descr->sorter(*m_table, index_pairs);
 
-        // Sorting can be specified by multiple columns, so that if two entries in the first column are
-        // identical, then the rows are ordered according to the second column, and so forth. For the
-        // first column, we cache all the payload of fields of the view in a std::vector<Mixed>
-        predicate.cache_first_column(index_pairs);
+        // Some descriptors, like Sort and Distinct, needs us to gather the current rows
+        // into a container we can use std algorithms on
+        if (base_descr->need_indexpair()) {
+            if (!using_indexpairs) {
+                use_indexpairs();
+            }
+
+            BaseDescriptor::Sorter predicate = base_descr->sorter(*m_table, index_pairs);
 
-        base_descr->execute(index_pairs, predicate, next);
+            // Sorting can be specified by multiple columns, so that if two entries in the first column are
+            // identical, then the rows are ordered according to the second column, and so forth. For the
+            // first column, we cache all the payload of fields of the view in a std::vector<Mixed>
+            predicate.cache_first_column(index_pairs);
+
+            base_descr->execute(index_pairs, predicate, next);
+        }
+        else {
+            if (using_indexpairs) {
+                apply_indexpairs();
+            }
+            base_descr->execute(*m_table, m_key_values, next);
```

Comment:
That's now quite a bit confusing since it's done under do_sort. Seems to me like original design not longer serves its purpose. Side effect from what i can tell that adding sort description into tableview will still call filter and stuff. How about  splitting this?